File: codex-aggregator/health_monitor.py
import requests
import time
import os

# Configuration
TOKENS = os.getenv("COPILOT_TOKENS", "").split(",")  # Comma-separated list of tokens
NOTIFICATION_URL = os.getenv("NOTIFICATION_URL")  # e.g., Bark or Telegram webhook
CHECK_INTERVAL = 3600  # Check every hour

def send_alert(message):
    if NOTIFICATION_URL:
        try:
            requests.get(f"{NOTIFICATION_URL}/{message}")
            print(f"Alert sent: {message}")
        except Exception as e:
            print(f"Failed to send alert: {e}")

def check_token(token):
    url = "https://api.github.com/user/billing/copilot"
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
    }
    try:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            return True, "Active"
        elif response.status_code == 422:
            return False, "Subscription Expired or Billing Issue"
        elif response.status_code == 401:
            return False, "Token Invalid or Revoked"
        else:
            return False, f"Unexpected Status: {response.status_code}"
    except Exception as e:
        return False, f"Error: {str(e)}"

def main():
    print("Starting Copilot Health Monitor...")
    while True:
        for token in TOKENS:
            if not token: continue
            is_active, status = check_token(token)
            if not is_active:
                msg = f"Codex Alert: Account [{token[:8]}...] status: {status}"
                send_alert(msg)
        time.sleep(CHECK_INTERVAL)

if __name__ == "__main__":
    main()

File: codex-aggregator/src/bin/health_monitor.rs
use anyhow::{Context, Result};
use clap::Parser;
use reqwest::blocking::Client;
use reqwest::StatusCode;
use std::env;
use std::thread;
use std::time::Duration;

#[derive(Parser)]
#[command(name = "codex-health-monitor")]
#[command(about = "Rust Copilot token health monitor for codex-aggregator")]
struct Args {
    #[arg(long)]
    once: bool,
    #[arg(long, default_value_t = 3600)]
    check_interval_seconds: u64,
}

#[derive(Debug)]
struct TokenStatus {
    active: bool,
    status: String,
}

fn main() -> Result<()> {
    let args = Args::parse();
    let client = Client::builder()
        .timeout(Duration::from_secs(20))
        .user_agent("codex-aggregator-health-monitor-rs/1.0")
        .build()
        .context("failed to build HTTP client")?;
    let tokens = env::var("COPILOT_TOKENS")
        .unwrap_or_default()
        .split(',')
        .map(str::trim)
        .filter(|token| !token.is_empty())
        .map(ToOwned::to_owned)
        .collect::<Vec<_>>();
    let notification_url = env::var("NOTIFICATION_URL")
        .ok()
        .filter(|value| !value.is_empty());

    println!("Starting Copilot Health Monitor...");
    loop {
        for token in &tokens {
            let status = check_token(&client, token);
            if !status.active {
                let message = format!(
                    "Codex Alert: Account [{}...] status: {}",
                    token.chars().take(8).collect::<String>(),
                    status.status
                );
                send_alert(&client, notification_url.as_deref(), &message);
            }
        }

        if args.once {
            break;
        }
        thread::sleep(Duration::from_secs(args.check_interval_seconds));
    }

    Ok(())
}

fn check_token(client: &Client, token: &str) -> TokenStatus {
    let request = client
        .get("https://api.github.com/user/billing/copilot")
        .bearer_auth(token)
        .header("Accept", "application/vnd.github+json");

    match request.send() {
        Ok(response) => match response.status() {
            StatusCode::OK => TokenStatus {
                active: true,
                status: "Active".to_string(),
            },
            StatusCode::UNPROCESSABLE_ENTITY => TokenStatus {
                active: false,
                status: "Subscription Expired or Billing Issue".to_string(),
            },
            StatusCode::UNAUTHORIZED => TokenStatus {
                active: false,
                status: "Token Invalid or Revoked".to_string(),
            },
            other => TokenStatus {
                active: false,
                status: format!("Unexpected Status: {}", other.as_u16()),
            },
        },
        Err(error) => TokenStatus {
            active: false,
            status: format!("Error: {error}"),
        },
    }
}

fn send_alert(client: &Client, notification_url: Option<&str>, message: &str) {
    let Some(base_url) = notification_url else {
        return;
    };
    let url = format!(
        "{}/{}",
        base_url.trim_end_matches('/'),
        urlencoding::encode(message)
    );
    match client.get(url).send() {
        Ok(_) => println!("Alert sent: {message}"),
        Err(error) => eprintln!("Failed to send alert: {error}"),
    }
}
